source/fem-dem/ib_particles_dem.cc

```diff
@@ -372,31 +382,72 @@ IBParticlesDEM<dim>::calculate_pw_contact_force(
   double wall_restitution_coefficient =
     parameters->wall_restitution_coefficient;
 
+  struct DefaultDBL_MAX
+  {
+    double value = DBL_MAX;
+  };
+  struct DefaultUINT_MAX
+  {
+    int value = UINT_MAX;
+  };
+  enum lowest_floating_wall_indices
+  {
+    lowest_floating_wall_indices = 1000000
+  };
+
+
   // Loop over the particles
   for (auto &particle : dem_particles)
     {
-      unsigned int boundary_index = 0;
-      double       best_dist      = DBL_MAX;
-      unsigned int best_index     = UINT_MAX;
+      unsigned int                              boundary_index = 0;
+      std::map<unsigned int, DefaultDBL_MAX>    best_dist;
+      std::map<unsigned int, DefaultUINT_MAX>   best_indices;
+      std::map<unsigned int, BoundaryCellsInfo> best_cells;
       // For each particle loop over the point and normal identified as
       // potential contact candidate.
       for (auto &boundary_cell_iter : boundary_cells[particle.particle_id])
         {
           // find the best candidate (the closest point).
           double dist =
             (boundary_cell_iter.point_on_boundary - particle.position).norm();
-          if (dist < best_dist)
+          if (dist < best_dist[boundary_cell_iter.boundary_index].value)
             {
-              best_dist  = dist;
-              best_index = boundary_index;
+              best_dist[boundary_cell_iter.boundary_index].value = dist;
+              best_indices[boundary_cell_iter.boundary_index].value =
+                boundary_index;
+              best_cells[boundary_cell_iter.boundary_index] =
+                boundary_cell_iter;
             }
           boundary_index += 1;
         }
+
+      // start the indexing of the floating wall at 1000000 so it doesn't
+
+      // add all the floating wall has contact candidates. Their indices start
+      // from 1000000 to avoid conflict with other boundary condition.
```

Comment:
```suggestion
      // We start the indexing of the floating wall at 1000000 so it does not
      // add all the floating wall as contact candidates. Their indices start
      // from lowest_floating_wall_index to avoid conflict with other boundary conditions.
```